// @sap/cds-compiler/lib/utils/timetrace.js
'use strict';

/**
 * A single StopWatch encapsulates the runtime of a selected code frame.
 *
 * @class TimeTrace
 */
class StopWatch {
  /**
   * Creates an instance of TimeTrace.
   * @param {string} id
   *
   * @memberOf TimeTrace
   */
  constructor(id) {
    this.id = id;
    this.startTime = process.hrtime.bigint();
    this.lapTime = this.startTime;
  }

  /**
   * Start watch.
   */
  start() {
    this.startTime = process.hrtime.bigint();
    this.lapTime = this.startTime;
  }

  /**
   * Stop and return delta T in nanoseconds,
   * but do not set start time
   */
  stop() {
    const endTime = process.hrtime.bigint();
    return endTime - this.startTime;
  }

  lap() {
    const endTime = process.hrtime.bigint();
    const dt = endTime - this.startTime;
    this.lapTime = process.hrtime.bigint();
    return dt;
  }

  stopInFloatSecs() {
    const dt = this.stop();
    return dt / BigInt(1000000000);
  }

  // lap as sec.ns float
  lapInFloatSecs() {
    const dt = this.lap();
    return dt / BigInt(1000000000);
  }
}

/**
 * The main class to handle measuring the runtime of code blocks
 *
 * Results are logged to stderr
 *
 * To enable time tracing, set CDSC_TIMETRACING to true in the environment
 *
 * @class TimeTracer
 */
class TimeTracer {
  /**
   * Creates an instance of TimeTracer.
   *
   * @memberOf TimeTracer
   */
  constructor() {
    this.traceStack = [];
  }

  /**
   * Start a new TimeTrace, using the given id for logging etc.
   *
   * @param {string} id A short description of whats going on
   *
   * @memberOf TimeTracer
   */
  start(id) {
    try {
      const b = new StopWatch(id);
      this.traceStack.push(b);
      b.start();
      // eslint-disable-next-line no-console
      console.error(`${ ' '.repeat((this.traceStack.length - 1) * 2) }${ id } started`);
    }
    catch (e) {
      // eslint-disable-next-line no-console
      console.error(`Starting time trace with id ${ id } failed: ${ e }`);
    }
  }

  /**
   * Stop the current TimeTrace and log the execution time.
   *
   *
   * @memberOf TimeTracer
   */
  stop() {
    try {
      const current = this.traceStack.pop();
      const dt = current.stop();
      const base = `${ ' '.repeat(this.traceStack.length * 2) }${ current.id } took:`;
      const sec = (dt / BigInt(1000000000)).toString();
      // first, get remaining ns, then convert to ms.
      const msec = ((dt % BigInt(1000000000)) / BigInt(1000000)).toString();
      // eslint-disable-next-line no-console
      console.error( `${ base }${ ' '.repeat(60 - base.length) } %ds %dms`, sec, msec );
    }
    catch (e) {
      // eslint-disable-next-line no-console
      console.error(`Stopping time trace failed: ${ e }`);
    }
  }
}

const ignoreTimeTrace = {
  start: () => { /* ignore */ },
  stop: () => { /* ignore */ },
};

const doTimeTrace = process && process.env && process.env.CDSC_TIMETRACING !== undefined;
module.exports = {
  timetrace: (doTimeTrace ? new TimeTracer() : ignoreTimeTrace),
  TimeTracer,
  StopWatch,
};
